File: Day_3/day_3.R
library(readxl)
library(dplyr)
library(ggplot2)

if (interactive() && requireNamespace("rstudioapi", quietly = TRUE)) {
  setwd(dirname(rstudioapi::getActiveDocumentContext()$path))
}
destfile <- "dataset_world_health_stat_2022.xlsx"
if (!file.exists(destfile)) {
  download.file("https://datasafe-h5afbhf4gwctabaa.z01.azurefd.net/api/Download/STR/BE2BEEE/2022-05-21/be2beee_20220520_0.xlsx", destfile, mode = "wb")
}

highlight_countries <- c("Bangladesh", "India", "Pakistan", "Sri Lanka", "Nepal", "Bhutan", "Maldives", "Afghanistan", "China", "United Kingdom", "United States", "Mexico", "Brazil", "South Africa", "Nigeria", "Australia")

df <- suppressWarnings(read_excel(destfile, sheet = "Annex 2-4", col_names = FALSE))
data <- df[6:nrow(df), c(1, 2, 6)]
colnames(data) <- c("Country", "Stunting", "Overweight")
data <- data %>%
  mutate(across(c(Stunting, Overweight), as.numeric)) %>%
  filter(!is.na(Stunting)) %>%
  mutate(
    ShortName = sapply(Country, function(c) {
      m <- highlight_countries[sapply(highlight_countries, function(n) grepl(n, c, ignore.case = TRUE))]
      if (length(m) > 0) m[1] else c
    }),
    Highlight = ShortName %in% highlight_countries
  ) %>%
  filter(Highlight) %>%
  arrange(Stunting) %>%
  mutate(ShortName = factor(ShortName, levels = ShortName))

plot <- ggplot(data) +
  geom_bar(aes(x = ShortName, y = Stunting, fill = "Stunting"), stat = "identity", color = "white") +
  geom_bar(aes(x = ShortName, y = -Overweight, fill = "Overweight"), stat = "identity", color = "white") +
  geom_text(aes(x = ShortName, y = Stunting, label = sprintf("%.1f%%", Stunting)), size = 3, hjust = -0.2, color = "#1A3D5C") +
  geom_text(aes(x = ShortName, y = -Overweight, label = sprintf("%.1f%%", Overweight)), size = 3, hjust = 1.2, color = "#1A3D5C") +
  scale_fill_manual(values = c("Stunting" = "#3D5A80", "Overweight" = "#E07A5F")) +
  coord_flip(ylim = c(-max(data$Overweight, na.rm = T) - 15, max(data$Stunting, na.rm = T) + 10)) +
  theme_minimal() +
  labs(title = "Childhood Malnutrition: Overweight vs Stunting", subtitle = "WHO World Health Statistics 2022  ·  % Under-5s", x = "", y = "", fill = "") +
  theme(
    plot.background = element_rect(fill = "#F7F3EE", color = NA), panel.background = element_rect(fill = "#F7F3EE", color = NA),
    panel.grid.major.x = element_line(color = "#E0D9D0", linetype = "dashed", linewidth = 0.8), panel.grid.minor = element_blank(), panel.grid.major.y = element_blank(),
    plot.title = element_text(color = "#1A1A2E", face = "bold", size = 18, margin = margin(b = 5)), plot.subtitle = element_text(color = "#1A1A2E", size = 11, margin = margin(b = 15)),
    axis.text.x = element_blank(),
    axis.text.y = element_text(color = sapply(levels(data$ShortName), function(s) if (s == "Bangladesh") "#006A4E" else "#1A3D5C"), face = sapply(levels(data$ShortName), function(s) if (s == "Bangladesh") "bold" else "plain")),
    legend.position = "bottom"
  )

ggsave("day_3_r.png", plot, width = 10, height = 8, dpi = 300, bg = "#F7F3EE")

File: Day_3/day_3.py
import os
import urllib.request
import pandas as pd
import matplotlib.pyplot as plt
import matplotlib.patheffects as pe

highlight_countries = ['Bangladesh', 'India', 'Pakistan', 'Sri Lanka', 'Nepal', 'Bhutan', 'Maldives', 'Afghanistan', 'China', 'United Kingdom', 'United States', 'Mexico', 'Brazil', 'South Africa', 'Nigeria', 'Australia']

PALETTE = {
    'background':   '#F7F3EE', 'grid': '#E0D9D0', 'bar_wasting': '#E07A5F', 'bar_stunting': '#3D5A80',
    'label_hl':     '#1A3D5C', 'label_bd': '#006A4E', 'title': '#1A1A2E', 'axis': '#4A4A5A',
}

def get_shortname(c_name):
    for k in highlight_countries:
        if k.lower() in str(c_name).lower(): return k
    return c_name

def main():
    script_dir = os.path.dirname(os.path.abspath(__file__))
    file_path = os.path.join(script_dir, 'dataset_world_health_stat_2022.xlsx')
    if not os.path.exists(file_path): urllib.request.urlretrieve('https://datasafe-h5afbhf4gwctabaa.z01.azurefd.net/api/Download/STR/BE2BEEE/2022-05-21/be2beee_20220520_0.xlsx', file_path)

    df = pd.read_excel(file_path, sheet_name='Annex 2-4', header=None)
    data = df.iloc[5:].copy()
    data = data[[0, 1, 3]]
    data.columns = ['Country', 'Stunting', 'Wasting']
    data['Stunting'] = pd.to_numeric(data['Stunting'], errors='coerce')
    data['Wasting'] = pd.to_numeric(data['Wasting'], errors='coerce')
    data = data.dropna()
    data['ShortName'] = data['Country'].apply(get_shortname)
    data['Highlight'] = data['ShortName'].isin(highlight_countries)
    data = data[data['Highlight']].sort_values('Stunting', ascending=True)

    fig, ax = plt.subplots(figsize=(10, 8))
    fig.patch.set_facecolor(PALETTE['background'])
    ax.set_facecolor(PALETTE['background'])
    ax.grid(True, axis='x', color=PALETTE['grid'], linewidth=0.8, linestyle='--', alpha=0.9)
    ax.set_axisbelow(True)

    bars_w = ax.barh(data['ShortName'], -data['Wasting'], color=PALETTE['bar_wasting'], label='Wasting (Acute)', edgecolor='white', linewidth=1)
    bars_s = ax.barh(data['ShortName'], data['Stunting'], color=PALETTE['bar_stunting'], label='Stunting (Chronic)', edgecolor='white', linewidth=1)
    ax.axvline(0, color=PALETTE['axis'], linewidth=1.5)

    for bar in bars_w:
        ax.text(bar.get_width() - 0.5, bar.get_y() + bar.get_height()/2, f"{-bar.get_width():.1f}%", ha='right', va='center', fontsize=9, color=PALETTE['label_hl'])
    for bar in bars_s:
        ax.text(bar.get_width() + 0.5, bar.get_y() + bar.get_height()/2, f"{bar.get_width():.1f}%", ha='left', va='center', fontsize=9, color=PALETTE['label_hl'])

    ax.set_xticks([])
    for spine in ax.spines.values(): spine.set_visible(False)
    for label in ax.get_yticklabels():
        if label.get_text() == 'Bangladesh':
            label.set_color(PALETTE['label_bd'])
            label.set_weight('bold')
        else: label.set_color(PALETTE['label_hl'])

    fig.text(0.1, 0.95, 'Childhood Malnutrition: Wasting vs Stunting', fontsize=17, fontweight='bold', color=PALETTE['title'], va='top')
    fig.text(0.1, 0.92, 'WHO World Health Statistics 2022  ·  % Under-5s', fontsize=10, color=PALETTE['title'], va='top', alpha=0.75)
    ax.legend(loc='lower right', frameon=True, facecolor=PALETTE['background'], edgecolor=PALETTE['grid'])
    fig.text(0.98, 0.015, '#30DayChartChallenge', ha='right', fontsize=9, color='#AAAAAA')
    plt.tight_layout(rect=[0, 0.02, 1, 0.90])
    plt.savefig(os.path.join(script_dir, 'day_3_python.png'), dpi=300, facecolor=PALETTE['background'], bbox_inches='tight')

if __name__ == '__main__': main()
